### PR TITLE
Destination snowflake: T+D via merge

### PR DESCRIPTION
**Pull request by edgao**
part of https://github.com/airbytehq/airbyte/issues/30764

The merge statement replaces three statements in the previous version:
* Insert new raw records to the final table
* Dedup the final table
* Execute CDC deletions

structurally, the merge is identical as https://github.com/airbytehq/airbyte/pull/31529. There are just some very minor syntax differences, plus I uppercased all the identifiers because ❄️ .

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`d3c830f178`](https://github.com/airbytehq/airbyte/pull/31683/commits/d3c830f178e218d99e35078e8693472cd1c48aeb)) - ❌

⏲️  Total pipeline duration: 07mn43s 

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Build connector tar                                                   | ✅       |
| Java Connector Unit Tests                                             | ✅       |
| Build destination-snowflake docker image for platform(s) linux/x86_64 | ✅       |
| Java Connector Integration Tests                                      | ✅       |
| Validate metadata for destination-snowflake                           | ✅       |
| Connector version semver check                                        | ✅       |
| Connector version increment check                                     | ❌       |
| QA checks                                                             | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_snowflake_td_via_merge/1697837688/d3c830f178e218d99e35078e8693472cd1c48aeb/destination-snowflake/3.2.3/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:d3c830f178e218d99e35078e8693472cd1c48aeb)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`3e00ac785c`](https://github.com/airbytehq/airbyte/pull/31683/commits/3e00ac785cd1451c1a80a79dabb32d2ce77640c1)) - ❌

⏲️  Total pipeline duration: 07mn42s 

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Build connector tar                                                   | ✅       |
| Java Connector Unit Tests                                             | ✅       |
| Build destination-snowflake docker image for platform(s) linux/x86_64 | ✅       |
| Java Connector Integration Tests                                      | ✅       |
| Validate metadata for destination-snowflake                           | ✅       |
| Connector version semver check                                        | ✅       |
| Connector version increment check                                     | ❌       |
| QA checks                                                             | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_snowflake_td_via_merge/1697838440/3e00ac785cd1451c1a80a79dabb32d2ce77640c1/destination-snowflake/3.2.3/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:3e00ac785cd1451c1a80a79dabb32d2ce77640c1)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGenerator.java [hunk 1]
```diff
@@ -315,6 +308,100 @@ String insertNewRecords(final StreamConfig stream,
...
+    final String cursorComparison;
+    if (stream.cursor().isPresent()) {
+      final String cursor = stream.cursor().get().name(QUOTE);
+      // Build a condition for "NEW_RECORD is more recent than TARGET_TABLE":
+      cursorComparison =
+          // First, compare the cursors.
+          "(TARGET_TABLE." + cursor + " < NEW_RECORD." + cursor
+          // Then, break ties with extracted_at. (also explicitly check for both NEW_RECORD and final table
+          // having null cursor
+          // because NULL != NULL in SQL)
```
Comment (jbfbell) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGenerator.java:338]: 😭 I have always despised this approach to null equality

**Review by jbfbell — APPROVED** [review]

**Review by cynthiaxyin — APPROVED** [review]
nit: I wonder if there's some beneficial refactoring we can do to pull out constants and the substituted SQL strings to reduce repeated code among V2 destinations. For example, for this change, it would be far simpler to simply set a boolean to indicate if I want `TARGET_TABLE` and `NEW_RECORD` to be upper or lower-cased, and the right syntax for the `MERGE` statement, and then the shared code would do the right thing to construct the full SQL statement (bc AFAICT those are the only two things different from the BQ implementation). And you could apply this in some degree to other parts of the file? 🤷‍♀️

**Comment by edgao** [comment]
> refactoring

maybe! This feels like a tradeoff between code reuse and readability - presumably doing that sort of conditional will make it a little harder to trace what the sqlgenerator actually does.

I wasn't sold on e.g. jooq back when we were starting this project, but cross-dialect translation is a problem that other people also run into - perhaps that's worth another look?

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`7bea287b55`](https://github.com/airbytehq/airbyte/pull/31683/commits/7bea287b5511260081e44bc92ce322b3841e1ab8)) - ❌

⏲️  Total pipeline duration: 12mn45s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Build connector tar                                                  | ✅       |
| Java Connector Unit Tests                                            | ✅       |
| Build destination-snowflake docker image for platform(s) linux/amd64 | ✅       |
| Java Connector Integration Tests                                     | ✅       |
| Validate metadata for destination-snowflake                          | ✅       |
| Connector version semver check                                       | ✅       |
| Connector version increment check                                    | ❌       |
| QA checks                                                            | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_snowflake_td_via_merge/1698253449/7bea287b5511260081e44bc92ce322b3841e1ab8/destination-snowflake/3.3.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:7bea287b5511260081e44bc92ce322b3841e1ab8)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by airbyte-oss-build-runner** [comment]
##  destination-snowflake test report (commit [`a35be661ee`](https://github.com/airbytehq/airbyte/pull/31683/commits/a35be661ee571fc89c0349a5474f9e563d3766c8)) - ❌

⏲️  Total pipeline duration: 15mn14s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Build connector tar                                                  | ✅       |
| Java Connector Unit Tests                                            | ✅       |
| Build destination-snowflake docker image for platform(s) linux/amd64 | ✅       |
| Java Connector Integration Tests                                     | ✅       |
| Validate metadata for destination-snowflake                          | ✅       |
| Connector version semver check                                       | ✅       |
| Connector version increment check                                    | ❌       |
| QA checks                                                            | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/edgao_snowflake_td_via_merge/1698343073/a35be661ee571fc89c0349a5474f9e563d3766c8/destination-snowflake/3.3.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:a35be661ee571fc89c0349a5474f9e563d3766c8)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-snowflake test
```

**Comment by sumeetgajjar** [comment]
Hi @edgao - First of all thank you for implementing the `MERGE` approach for Snowflake.

I see that the commit was reverted. Can you please elaborate on the reason for the revert?

We were facing higher execution times with the `T + D` approach similar to https://github.com/airbytehq/airbyte/issues/43046 and our experiment shows that the `MERGE` approach would reduce the execution time drastically.

**Comment by gisripa** [comment]
@sumeetgajjar  We reverted because of users reporting increased warehouse activity. We did not have empirical proof at that time to prove otherwise so we reverted to old behavior.

**Pull request merged**